src/util/mpir_hwtopo.c

```diff
@@ -202,22 +202,31 @@ int MPII_hwtopo_init(void)
 
 #ifdef HAVE_HWLOC
     bindset = hwloc_bitmap_alloc();
-    hwloc_topology_init(&hwloc_topology);
-    char *xmlfile = MPIR_pmi_get_jobattr("PMI_hwloc_xmlfile");
-    if (xmlfile != NULL) {
-        int rc;
-        rc = hwloc_topology_set_xml(hwloc_topology, xmlfile);
-        if (rc == 0) {
-            /* To have hwloc still actually call OS-specific hooks, the
-             * HWLOC_TOPOLOGY_FLAG_IS_THISSYSTEM has to be set to assert that the loaded
-             * file is really the underlying system. */
-            hwloc_topology_set_flags(hwloc_topology, HWLOC_TOPOLOGY_FLAG_IS_THISSYSTEM);
+
+    mpi_errno = MPIR_pmi_load_hwloc_topology(&hwloc_topology);
+    if (MPI_SUCCESS == mpi_errno) {
+        from_pmi = 1;
+    } else {
+        hwloc_topology_init(&hwloc_topology);
+        char *xmlfile = MPIR_pmi_get_jobattr("PMI_hwloc_xmlfile");
+        if (xmlfile != NULL) {
+            int rc;
+            rc = hwloc_topology_set_xml(hwloc_topology, xmlfile);
+            if (rc == 0) {
+                /* To have hwloc still actually call OS-specific hooks, the
+                 * HWLOC_TOPOLOGY_FLAG_IS_THISSYSTEM has to be set to assert that the loaded
+                 * file is really the underlying system. */
+                hwloc_topology_set_flags(hwloc_topology, HWLOC_TOPOLOGY_FLAG_IS_THISSYSTEM);
+            }
+            MPL_free(xmlfile);
         }
```

Comment:
What I meant is to move this `else` branch entirely into `MPIR_pmi_load_hwloc_topology`. This else-branch effectively is a fallback implementation of `PMIx_load_topology`,

Reply:
I am not sure about that. The else branch is probably for hydra, not PMIx fallback.

Reply:
What I meant was that it is not a fallback to PMIx. It might be a fallback to pmi1_load_hwloc_topology or pmi2_load_hwloc_topology. if put else branch to PMIx, you don't get this path if not using PMIx but Hydra.